### PR TITLE
VE-3102: DMH vector search: AION updates

### PR DESCRIPTION
**Pull request by ndthang15**
# Overview
AION needs to be updated with additional properties that could be used by GLC and DMH vector search.

## Description
1. Added `referenceId` property to `objectResult` and `seriesItem` definitions.
2. Added `embedding` object (with `vector`, `tags`, and `referenceId` properties) to a new top-level in the AION schema.
3. Added `tags` property to the top level of `objectResult`.

```
{
    "series": [{
        "start": 0,
        "stop": 1000,
        "referenceId": "ABC" // this is a new property added to the seriesItem definition.
    },
    {
        "start": 2000,
        "stop": 3000,
        "referenceId": "ABC"
    }],
    "embedding": [{  // "embedding" is on the same level as "object" and "series" in the AION schema.
        "vector": [0.0, 0.1]
        "tags": [],
         "referenceId": "ABC"
    },
    {
        "vector": [2.0, 2.1]
        "tags": [],
         "referenceId": "ABC"
    }],
}
```

4. Updated AION schema with GLC prototype.
```
{
    "sourceEngineId": "AAAAAAAA-AAAA-AAAA-AAAA-AAAAAAAAAAAA", // existing property in the PREAMBLE definition.
    "internalTaskId": "eb20f1b3-20b4-471d-8dd9-76e077969af4", // this is a new property added to the PREAMBLE definition.
    "generatedDateUTC": "2024-05-15T15:27:33.848266649Z",
    "validationContracts": [
        "text"
    ],
	"object" : 
	[
		{
			"fingerprintVector" : [ // this is a new property added to the objectResult definition.
				// ...
			],
			"label" : "person",
			// this used to link timeSeries to a vector
			"referenceId" : "f46f029a-d42c-411c-ad01-8cffe3f87a63", // this is a new property added to the objectResult definition. 
			"tags" :       //  this is a new property added to the objectResult definition.
			[
				{
					"key" : "Accessory",
					"value" : "BagAny"
				}				
			],
			"type" : "fingerprint", //  existing property in the objectResult definition. 
			"vendor" : 
			{
				"label" : "fe_person_reid_512_v17_0_m34"
			}
		}
	],
	"series" : 
	[
		{
			"object" : 
			{
				"boundingPoly" : [
					// ...
				],
				"label" : "",
				"referenceId" : "f46f029a-d42c-411c-ad01-8cffe3f87a63", // this is a new property added to the objectResult definition.
				"type" : "object"
			},
			"startTimeMs" : 0,
			"stopTimeMs" : 0
		},
		{
			"object" : 
			{
				"boundingPoly" : [
					// ...
				],
				"label" : "",				
				"referenceId" : "f46f029a-d42c-411c-ad01-8cffe3f87a63", // this is a new property added to the objectResult definition.
				"type" : "object"
			},
			"startTimeMs" : 40,
			"stopTimeMs" : 40
		}				
	]
}
```
5. Added valid and invalid tests to AION schema.

## Related Issue

https://veritone.atlassian.net/browse/VE-3102

## How Has This Been Tested
1. Run `npm i && npm run test` in /packages/veritone-json-schemas folder.
2. Ensure that all tests are passed.

### CODE REVIEW

**Comment by mgiasiVeri** [comment]
@ndthang15 we also need to support GLC's new AION format which is similar but introduces a few more fields like fingerprintVector. The ticket links to an example  in the search draft doc, but the example is collapsed so you have to expand it. 

Open up https://veritone.atlassian.net/wiki/spaces/ENG/pages/3330900143/Vector+Search+Draft#AION-updates and search for "GLC prototype" to expand and see it. Note that the object object does not need to have start and stop times. Here is another view of GLCs needed AION changes. https://veritone.atlassian.net/wiki/spaces/VP/pages/2859532299/Next+generation+tracker+AION+format#Accepted-Solution

**Comment by ndthang15** [comment]
> @ndthang15 we also need to support GLC's new AION format which is similar but introduces a few more fields like fingerprintVector. The ticket links to an example in the search draft doc, but the example is collapsed so you have to expand it.
> 
> Open up https://veritone.atlassian.net/wiki/spaces/ENG/pages/3330900143/Vector+Search+Draft#AION-updates and search for "GLC prototype" to expand and see it. Note that the object object does not need to have start and stop times. Here is another view of GLCs needed AION changes. https://veritone.atlassian.net/wiki/spaces/VP/pages/2859532299/Next+generation+tracker+AION+format#Accepted-Solution

@mgiasiVeri Today I have updated the schema to support GLC's new AION format base on your information. And this PR is ready for review again. Thank you so much for reviewing!

**Comment by mgiasiVeri** [comment]
@ndthang15 i think we need "tags" added at the objectResult level too. If you look at the GLC example, they leverage tags in the top level object, which is outside of the seriesItem. Let me know if that isn't clear.

**Comment by ndthang15** [comment]
> @ndthang15 i think we need "tags" added at the objectResult level too. If you look at the GLC example, they leverage tags in the top level object, which is outside of the seriesItem. Let me know if you have any questions. And to be clear, we do not need the start and stop time values added at the objectResult level, that is a mistake in the examples provided.

@mgiasiVeri You're right that I'm missing the `tags` in `objectResult`, so I've just added it in the latest commit.
About the start and stop time values at the `objectResult` level, we didn't add it to the schema. It is also a mistake when I copied the sample schema in document to PR's description.
Thank you so much for your notes!

**Comment by crondonveritone** [comment]
Based on the example of GLC prototype https://veritone.atlassian.net/wiki/spaces/ENG/pages/3330900143/Vector+Search+Draft#GLC-prototype%3A

I see that fingerPrintVector has the properties `label`, `referenceId`, and `type` but in the example, these properties are outer of fingerPrintVector (at _#/definitions/objectResult_ level). If the second case is correct then it LGTM:

**Comment by mgiasiVeri** [comment]
> Based on the information of GLC prototype https://veritone.atlassian.net/wiki/spaces/ENG/pages/3330900143/Vector+Search+Draft#GLC-prototype%3A
> 
> I see that fingerPrintVector has the properties `label`, `referenceId`, and `type` but in the example, these properties are outer of fingerPrintVector (at _#/definitions/objectResult_ level). If the second case is correct then that part LGTM:

@crondonveritone I think that was a typo. label, referenced, and type should all be outside of fingerprintVector. And fingerprintVector should only be an array of numbers. @frankayars @alex-oleksiiuk could you confirm.

File: packages/veritone-json-schemas/schemas/vtn-standard/master.json [hunk 1]
```diff
@@ -767,6 +785,30 @@
     "vendor": {
       "description": "Custom data that doesn't conform to any other field. You can add any arbitrary data inside this object, but it will not be indexed, searchable, or have any impact on the system. However it will be returned when reading the data back out.",
       "type": "object"
+    },
+    "fingerprintVector": {
```
[the remark this answers is not in the document]
Comment (ndthang15) [reply packages/veritone-json-schemas/schemas/vtn-standard/master.json:789]: I defined the type of items in the fingerprintVector array. I think we don't need to define range for it, [like the KNN example in the ES document](https://www.elastic.co/guide/en/elasticsearch/reference/current/knn-search.html#approximate-knn):
```
curl -X POST "localhost:9200/image-index/_bulk?refresh=true&pretty" -H 'Content-Type: application/json' -d'
{ "index": { "_id": "1" } }
{ "image-vector": [1, 5, -20], "title-vector": [12, 50, -10, 0, 1], "title": "moose family", "file-type": "jpg" }
{ "index": { "_id": "2" } }
{ "image-vector": [42, 8, -15], "title-vector": [25, 1, 4, -12, 2], "title": "alpine lake", "file-type": "png" }
{ "index": { "_id": "3" } }
{ "image-vector": [15, 11, 23], "title-vector": [1, 5, 25, 50, 20], "title": "full moon", "file-type": "jpg" }
...
'
```
 Could you please review again? Thanks you!

**Review by crondonveritone — APPROVED** [review]
LGTM

File: packages/veritone-json-schemas/schemas/vtn-standard/master.json [hunk 2]
```diff
@@ -767,6 +785,33 @@
...
+      "description": "An array of vectors related to the vectorized data",
```
Comment (frankayars) [inline packages/veritone-json-schemas/schemas/vtn-standard/master.json:790]: Suggestion: 
An array of floats representing objects in vector space.

File: packages/veritone-json-schemas/schemas/vtn-standard/master.json [hunk 3]
```diff
@@ -767,6 +785,33 @@
...
+      "type": "array",
+      "items": {
+        "type": "number"
+      }
+    },
+    "embedding": {
+      "description": "The embedding engine result was generated",
```
Comment (frankayars) [inline packages/veritone-json-schemas/schemas/vtn-standard/master.json:797]: Suggestion: 
An array of floats representing objects in vector space.

**Review by frankayars — COMMENTED** [review]
Suggested two minor description changes.

**Comment by ndthang15** [comment]
> Suggested two minor description changes.

@frankayars Thanks for your suggestions. I changed the descriptions of these definitions. Also, the fingerprintVector definition was renamed 'vector' to a more generic name. Could you please review again?

**Review by mgiasiVeri — APPROVED** [review]

**Pull request merged**